File: taurus-protect-sdk-typescript/src/models/transaction.ts
/**
 * Transaction domain models for Taurus-PROTECT SDK.
 */

import type { OffsetPageOptions } from "./pagination";

/**
 * Transaction status enum.
 */
export enum TransactionStatus {
  NEW = "NEW",
  SUCCESS = "SUCCESS",
  PENDING = "PENDING",
  TEMPORARY_FAILURE = "TEMPORARY_FAILURE",
  INVALID = "INVALID",
  TIMEOUT = "TIMEOUT",
  MINED = "MINED",
  LOST = "LOST",
  EXPIRED = "EXPIRED",
}

/**
 * Transaction direction enum.
 */
export enum TransactionDirection {
  INCOMING = "incoming",
  OUTGOING = "outgoing",
}

/**
 * Transaction type enum.
 */
export enum TransactionType {
  TRANSFER = "transfer",
  BURN = "burn",
  APPROVE = "approve",
  STAKE = "stake",
  UNSTAKE = "unstake",
}

/**
 * Represents address information within a transaction.
 */
export interface TransactionAddressInfo {
  /** The blockchain address */
  readonly address: string | undefined;
  /** Label of the address if known to Taurus-PROTECT */
  readonly label: string | undefined;
  /** Container identifier */
  readonly container: string | undefined;
  /** Customer identifier */
  readonly customerId: string | undefined;
  /** Amount in smallest currency unit */
  readonly amount: string | undefined;
  /** Amount in main currency unit */
  readonly amountMainUnit: string | undefined;
  /** Type of the transaction element (source or destination) */
  readonly type: string | undefined;
  /** Index within transaction */
  readonly idx: string | undefined;
  /** Internal address ID if this is an internal address */
  readonly internalAddressId: string | undefined;
  /** Whitelisted address ID if this is a whitelisted address */
  readonly whitelistedAddressId: string | undefined;
}

/**
 * Currency information for a transaction.
 */
export interface TransactionCurrencyInfo {
  /** Currency ID */
  readonly id: string | undefined;
  /** Currency symbol (e.g., "ETH", "BTC") */
  readonly symbol: string | undefined;
  /** Currency name */
  readonly name: string | undefined;
  /** Number of decimal places */
  readonly decimals: number | undefined;
  /** Blockchain identifier */
  readonly blockchain: string | undefined;
  /** Network identifier */
  readonly network: string | undefined;
}

/**
 * Transaction attribute (key-value metadata).
 */
export interface TransactionAttribute {
  /** Attribute key */
  readonly key: string | undefined;
  /** Attribute value */
  readonly value: string | undefined;
}

/**
 * Represents a blockchain transaction in Taurus-PROTECT.
 *
 * Transactions represent the movement of cryptocurrency on the blockchain
 * and can be either incoming (received) or outgoing (sent).
 */
export interface Transaction {
  /** Unique transaction identifier */
  readonly id: string;
  /** Transaction direction (incoming or outgoing) */
  readonly direction: TransactionDirection | string | undefined;
  /** Currency symbol or ID */
  readonly currency: string | undefined;
  /** Currency information */
  readonly currencyInfo: TransactionCurrencyInfo | undefined;
  /** Blockchain identifier (e.g., "ETH", "BTC") */
  readonly blockchain: string | undefined;
  /** Network identifier (e.g., "mainnet", "testnet") */
  readonly network: string | undefined;
  /** Blockchain transaction hash */
  readonly hash: string | undefined;
  /** Block number */
  readonly block: string | undefined;
  /** Confirmation block number */
  readonly confirmationBlock: string | undefined;
  /** Amount in smallest currency unit */
  readonly amount: string | undefined;
  /** Amount in main currency unit */
  readonly amountMainUnit: string | undefined;
  /** Fee in smallest currency unit */
  readonly fee: string | undefined;
  /** Fee in main currency unit */
  readonly feeMainUnit: string | undefined;
  /** Transaction type (e.g., "burn", "approve") */
  readonly type: string | undefined;
  /** Current transaction status */
  readonly status: TransactionStatus | string | undefined;
  /** Whether the transaction is fully confirmed */
  readonly isConfirmed: boolean;
  /** Source addresses */
  readonly sources: TransactionAddressInfo[];
  /** Destination addresses */
  readonly destinations: TransactionAddressInfo[];
  /** Transaction ID (internal) */
  readonly transactionId: string | undefined;
  /** Unique ID */
  readonly uniqueId: string | undefined;
  /** Associated request ID */
  readonly requestId: string | undefined;
  /** Whether the associated request is visible */
  readonly requestVisible: boolean | undefined;
  /** When the transaction was received by the blockchain */
  readonly receptionDate: Date | undefined;
  /** When the transaction was confirmed */
  readonly confirmationDate: Date | undefined;
  /** Transaction attributes */
  readonly attributes: TransactionAttribute[];
  /** Fork number (typically 0, 1 for new platforms like DOT AssetHub) */
  readonly forkNumber: string | undefined;
}

/**
 * Options for listing transactions. An offset list: `limit` 1-100 (default 20) and
 * `offset` (a previous page's `pagination.nextOffset`).
 */
export interface ListTransactionsOptions extends OffsetPageOptions {
  /** Filter by currency ID or symbol */
  readonly currency?: string;
  /** Filter by direction ("incoming" or "outgoing") */
  readonly direction?: TransactionDirection | string;
  /** Filter by blockchain */
  readonly blockchain?: string;
  /** Filter by network */
  readonly network?: string;
  /** Filter transactions after this date */
  readonly fromDate?: Date;
  /** Filter transactions before this date */
  readonly toDate?: Date;
  /** Search query string */
  readonly query?: string;
  /** Filter by transaction type */
  readonly type?: string;
  /** Filter by source address */
  readonly source?: string;
  /** Filter by destination address */
  readonly destination?: string;
  /** Filter by address (either source or destination) */
  readonly address?: string;
  /** Filter by transaction hashes */
  readonly hashes?: string[];
  /** Filter by transaction IDs */
  readonly ids?: string[];
  /** Filter by from block number */
  readonly fromBlockNumber?: string;
  /** Filter by to block number */
  readonly toBlockNumber?: string;
  /** Filter by amount above this value */
  readonly amountAbove?: string;
  /** Exclude transactions with unknown source/destination */
  readonly excludeUnknownSourceDestination?: boolean;
  /** Filter by customer ID */
  readonly customerId?: string;
}

/**
 * Options for exporting transactions.
 *
 * The export cannot page: the server always exports from the first matching row, so there
 * is no offset. `limit` (default 20, no SDK maximum) is the only way to reach more rows;
 * compare the result's `totalItems` with the rows exported to spot a truncated export.
 */
export interface ExportTransactionsOptions {
  /** How many transactions to export (default 20) */
  readonly limit?: number;
  /** Export format: "json" (the server default), "csv" or "csv_simple" */
  readonly format?: string;
  /** Filter transactions after this date */
  readonly fromDate?: Date;
  /** Filter transactions before this date */
  readonly toDate?: Date;
  /** Filter by currency ID or symbol */
  readonly currency?: string;
  /** Filter by direction ("incoming" or "outgoing") */
  readonly direction?: TransactionDirection | string;
  /** Filter by blockchain */
  readonly blockchain?: string;
  /** Filter by network */
  readonly network?: string;
  /** Search query string */
  readonly query?: string;
  /** Filter by address (either source or destination) */
  readonly address?: string;
}

/**
 * The result of a transaction export.
 */
export interface ExportTransactionsResult {
  /** The export payload, in the requested format ("" when nothing matched) */
  readonly data: string;
  /** How many transactions match the filters (0 when omitted); more than exported means truncated */
  readonly totalItems: number;
}

File: taurus-protect-sdk-go/pkg/protect/model/transaction.go
package model

import "time"

// Transaction represents a blockchain transaction.
type Transaction struct {
	// ID is the unique identifier for the transaction.
	ID string `json:"id"`
	// UniqueID is another unique identifier for the transaction.
	UniqueID string `json:"unique_id,omitempty"`
	// TransactionID is an internal transaction ID.
	TransactionID string `json:"transaction_id,omitempty"`
	// Hash is the blockchain transaction hash.
	Hash string `json:"hash,omitempty"`
	// Direction is the transaction direction (incoming, outgoing).
	Direction string `json:"direction"`
	// Type is the transaction type (transfer, burn, approve, etc.).
	Type string `json:"type,omitempty"`
	// Currency is the currency symbol.
	Currency string `json:"currency"`
	// Blockchain is the blockchain name (ETH, BTC, etc.).
	Blockchain string `json:"blockchain,omitempty"`
	// Amount is the transaction amount in smallest currency unit.
	Amount string `json:"amount"`
	// AmountMainUnit is the amount in main currency unit.
	AmountMainUnit string `json:"amount_main_unit,omitempty"`
	// Fee is the transaction fee in smallest currency unit.
	Fee string `json:"fee,omitempty"`
	// FeeMainUnit is the fee in main currency unit.
	FeeMainUnit string `json:"fee_main_unit,omitempty"`
	// Block is the block number.
	Block string `json:"block,omitempty"`
	// Sources are the sending addresses.
	Sources []TransactionAddressInfo `json:"sources,omitempty"`
	// Destinations are the receiving addresses.
	Destinations []TransactionAddressInfo `json:"destinations,omitempty"`
	// ReceptionDate is when the transaction was received by the blockchain.
	ReceptionDate time.Time `json:"reception_date,omitempty"`
	// ConfirmationDate is when the transaction was fully confirmed.
	ConfirmationDate time.Time `json:"confirmation_date,omitempty"`
}

// TransactionAddressInfo represents an address involved in a transaction.
type TransactionAddressInfo struct {
	// Address is the blockchain address string.
	Address string `json:"address"`
	// Label is the address label if known.
	Label string `json:"label,omitempty"`
	// CustomerID is the customer identifier if set.
	CustomerID string `json:"customer_id,omitempty"`
	// Amount is the transaction amount for this address.
	Amount string `json:"amount,omitempty"`
	// AmountMainUnit is the amount in main currency unit.
	AmountMainUnit string `json:"amount_main_unit,omitempty"`
	// Type is the element type (source or destination).
	Type string `json:"type,omitempty"`
	// InternalAddressID is the ID if this is an internal address.
	InternalAddressID string `json:"internal_address_id,omitempty"`
	// WhitelistedAddressID is the ID if this is a whitelisted address.
	WhitelistedAddressID string `json:"whitelisted_address_id,omitempty"`
}

// ListTransactionsOptions contains options for listing transactions.
type ListTransactionsOptions struct {
	// Limit is the page size: 0 selects DefaultPageSize, above MaxPageSize is an error.
	Limit int64
	// Offset is the number of transactions to skip; continue with Pagination.NextOffset.
	Offset int64
	// Currency filters by currency symbol.
	Currency string
	// Direction filters by transaction direction (incoming, outgoing).
	Direction string
	// Blockchain filters by blockchain.
	Blockchain string
	// Network filters by network.
	Network string
	// Query is a partial, case-insensitive match across address, label, customer id,
	// transaction id, hash and type. Prefer IDs/Hashes for an exact value.
	Query string
	// FromDate filters by transaction creation date (inclusive lower bound).
	FromDate *time.Time
	// ToDate filters by transaction creation date (inclusive upper bound).
	ToDate *time.Time
	// IDs filters to specific Taurus transaction IDs.
	IDs []string
	// Hashes filters to specific on-chain transaction hashes.
	Hashes []string
	// TransactionIDs filters by the chain-level transaction identifier.
	TransactionIDs []string
	// Address filters to transactions touching one blockchain address.
	Address string
	// Source filters by originating address.
	Source string
	// Destination filters by destination address.
	Destination string
	// AmountAbove keeps only transactions above this amount.
	AmountAbove string
}

// ListTransactionsByAddressOptions contains options for listing transactions by address.
type ListTransactionsByAddressOptions struct {
	// Limit is the page size: 0 selects DefaultPageSize, above MaxPageSize is an error.
	Limit int64
	// Offset is the number of transactions to skip; continue with Pagination.NextOffset.
	Offset int64
	// Currency filters by currency symbol.
	Currency string
	// Direction filters by transaction direction (incoming, outgoing).
	Direction string
	// Blockchain filters by blockchain.
	Blockchain string
}

// ExportTransactionsOptions contains options for exporting transactions.
type ExportTransactionsOptions struct {
	// From filters transactions after this date.
	From *time.Time
	// To filters transactions before this date.
	To *time.Time
	// Currency filters by currency ID or symbol.
	Currency string
	// Direction filters by transaction direction (incoming, outgoing).
	Direction string
	// Blockchain filters by blockchain.
	Blockchain string
	// Format specifies the export format: json (the server's default when empty), csv or
	// csv_simple (one line per transfer rather than per transaction).
	Format string
	// Limit is the number of transactions to export: 0 selects DefaultPageSize; there is no
	// SDK maximum. The export cannot page (the server always exports from the first matching
	// transaction), so a larger Limit is the only way to reach more rows.
	Limit int64
	// Address filters transactions involving this address.
	Address string
	// Query searches transaction fields.
	Query string
}

// ExportTransactionsResult is an export of transactions.
type ExportTransactionsResult struct {
	// Data is the export in the requested format.
	Data string `json:"data"`
	// TotalItems is the number of matching transactions; above Limit, the export is truncated.
	TotalItems int64 `json:"total_items"`
}
